src/scope.rs

```diff
@@ -1,292 +1,961 @@
+      (init_fn)(&mut *data_box);
+      let data_ptr = Box::into_raw(data_box);
+      isolate.set_current_scope(NonNull::new(data_ptr));
+      unsafe { &mut *data_ptr }
+    }
+
+    fn drop(&mut self) {
+      // Make our parent scope 'current' again.
+      let parent = self.parent;
+      let isolate = self.get_isolate_mut();
+      isolate.set_current_scope(parent);
+
+      // Turn the &mut self pointer back into a box and drop it.
+      let _ = unsafe { Box::from_raw(self) };
+    }
+
+    pub(super) fn notify_scope_dropped(&mut self) {
+      // This function is called when the `api::Scope` object is dropped.
+      // With regard to (escapable) handle scopes: the Rust borrow checker
+      // allows these to be dropped before all the local handles that were
+      // created inside the HandleScope have gone out of scope. In order to
+      // avoid turning these locals into invalid references the HandleScope is
+      // kept alive for now -- it'll be actually dropped when the user touches
+      // the HandleScope's parent scope.
+      match &self.type_specific_data {
+        ScopeTypeSpecificData::HandleScope(_)
+        | ScopeTypeSpecificData::EscapableHandleScope(_) => {
+          // Defer drop.
+          let prev_flag_value = replace(&mut self.deferred_drop, true);
+          assert_eq!(prev_flag_value, false);
+        }
+        _ => {
+          // Regular, immediate drop.
+          self.drop();
+        }
+      }
+    }
+
+    pub(super) fn as_scope<S: api::Scope>(&mut self) -> S {
+      assert_eq!(size_of::<&mut Self>(), size_of::<S>());
+      let self_nn = NonNull::from(self);
+      unsafe { ptr::read(&self_nn as *const _ as *const S) }
+    }
+
+    pub(super) fn get<S: api::Scope>(scope: &S) -> &Self {
+      let self_nn = unsafe { *(scope as *const _ as *const NonNull<Self>) };
+      Self::touch(self_nn);
+      unsafe { &*self_nn.as_ptr() }
+    }
+
+    pub(super) fn get_mut<S: api::Scope>(scope: &mut S) -> &mut Self {
+      let self_nn = unsafe { *(scope as *mut _ as *mut NonNull<Self>) };
+      Self::touch(self_nn);
+      unsafe { &mut *self_nn.as_ptr() }
+    }
+
+    pub(super) fn get_isolate(&self) -> &Isolate {
+      unsafe { self.isolate.as_ref() }
+    }
+
+    pub(super) fn get_isolate_mut(&mut self) -> &mut Isolate {
+      unsafe { self.isolate.as_mut() }
+    }
+
+    pub(super) fn get_isolate_ptr(&self) -> *mut Isolate {
+      self.isolate.as_ptr()
+    }
+
+    fn touch(self_nn: NonNull<ScopeData>) {
+      loop {
+        let current_scope = unsafe { self_nn.as_ref() }
+          .get_isolate()
+          .get_current_scope();
+        match current_scope {
+          Some(current_scope_nn) if current_scope_nn == self_nn => break,
+          Some(mut current_scope_nn)
+            if unsafe { current_scope_nn.as_ref().deferred_drop } =>
+          unsafe { current_scope_nn.as_mut().drop() }
+          _ => panic!("can't use scope that has been shadowed by another"),
+        }
+      }
```

Comment:
I guess we cannot claim this is a zero-overhead binding anymore...

Reply:
The loop iterates 0 or 1 times